src/circuit.rs

```diff
@@ -50,13 +270,401 @@ impl plonk::Circuit<pallas::Base> for Circuit {
+            //
+            let scalar = {
+                let scalar_val = nk_rho_old
+                    .value()
+                    .zip(psi_old.value())
+                    .map(|(nk_rho_old, psi_old)| nk_rho_old + psi_old);
+                let scalar = self.load_private(
+                    layouter.namespace(|| "poseidon_hash(nk, rho_old) + psi_old"),
+                    config.advices[0],
+                    scalar_val,
+                )?;
+
+                config.plonk_chip().add(
+                    layouter.namespace(|| "poseidon_hash(nk, rho_old) + psi_old"),
+                    nk_rho_old,
+                    psi_old,
+                    scalar,
+                    Some(pallas::Base::one()),
+                    Some(pallas::Base::one()),
+                    Some(pallas::Base::one()),
+                )?;
+
+                scalar
+            };
+
+            // Multiply scalar by NullifierK
+            // `product` = [poseidon_hash(nk, rho_old) + psi_old] NullifierK.
+            //
+            let product = {
+                let nullifier_k = FixedPointBaseField::from_inner(ecc_chip.clone(), NullifierK);
+                nullifier_k.mul(
+                    layouter.namespace(|| "[poseidon_output + psi_old] NullifierK"),
+                    scalar,
+                )?
+            };
+
+            // Add cm_old to multiplied fixed base to get nf_old
+            // cm_old + [poseidon_output + psi_old] NullifierK
+            cm_old
+                .add(layouter.namespace(|| "nf_old"), &product)?
+                .extract_p()
+        };
+
+        // Spend authority
+        // TODO: constrain to equal public input rk
+        let _rk = {
+            // alpha_commitment = [alpha] SpendAuthG
+            let (alpha_commitment, _) = {
+                let spend_auth_g = OrchardFixedBasesFull::SpendAuthG;
+                let spend_auth_g = FixedPoint::from_inner(ecc_chip.clone(), spend_auth_g);
+                spend_auth_g.mul(layouter.namespace(|| "[alpha] SpendAuthG"), self.alpha)?
+            };
+
+            // [alpha] SpendAuthG + ak
+            alpha_commitment.add(layouter.namespace(|| "rk"), &ak)?
+        };
+
+        // Diversified address integrity.
+        let (pk_d_old, _) = {
+            let commit_ivk_config = config.commit_ivk_config.clone();
+
+            let ivk = {
+                let rivk = self.rivk.map(|rivk| *rivk);
+
+                commit_ivk_config.assign_region(
+                    config.sinsemilla_chip_1(),
+                    ecc_chip.clone(),
+                    layouter.namespace(|| "CommitIvk"),
+                    *ak.extract_p().inner(),
+                    nk,
+                    rivk,
+                )?
+            };
+
+            // [ivk] g_d_old
+            g_d_old.mul(layouter.namespace(|| "[ivk] g_d_old"), ivk.inner())?
+        };
+
+        // Old note commitment integrity.
+        let _cm_old = {
```

Comment:
`_cm_old` here has to be constrained equal to `cm_old` used in the Merkle path check.

Reply:
This has now been implemented:
```
derived_cm_old.constrain_equal(layouter.namespace(|| "cm_old equality"), &cm_old)?;
```